src/charm.py

```diff
@@ -0,0 +1,257 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical
+# See LICENSE file for licensing details.
+
+"""Charm the calibre-web application."""
+
+import logging
+import typing
+from pathlib import Path
+from typing import cast
+
+import ops
+
+CONTAINER_NAME = 'calibre-web'
+SERVICE_NAME = 'calibre-web'
+STORAGE_NAME = 'books'
+LIBRARY_WRITE_ACTION = 'library-write'
+LIBRARY_WRITE_CONFIG = 'library-write'
+LIBRARY_INFO_ACTION = 'library-info'
+LIBRARY_INFO_FORMAT_PARAM = 'format'
+
+LibraryInfoFormat: typing.TypeAlias = typing.Literal['tree', 'ls-1']
+LIBRARY_INFO_FORMATS = typing.get_args(LibraryInfoFormat)
+
+LibraryWriteBehaviour: typing.TypeAlias = typing.Literal['skip', 'clean']
+LIBRARY_WRITE_BEHAVIOURS = typing.get_args(LibraryWriteBehaviour)
+
+
+CalibreWebLayerDict = typing.TypedDict(
+    'CalibreWebLayerDict',
+    {
+        'summary': str,
+        'description': str,
+        'services': dict[str, ops.pebble.ServiceDict],
+    },
+    total=True,
+)
+
+
+class CalibreWebCharm(ops.CharmBase):
+    """Charm the calibre-web application."""
+
+    def __init__(self, framework: ops.Framework) -> None:
+        super().__init__(framework)
+        framework.observe(self.on.collect_unit_status, self._on_collect_status)
+        framework.observe(self.on[CONTAINER_NAME].pebble_ready, self._on_pebble_ready)
+        framework.observe(self.on[STORAGE_NAME].storage_attached, self._on_storage_attached)
+        framework.observe(self.on.config_changed, self._on_config_changed)
+        framework.observe(self.on[LIBRARY_WRITE_ACTION].action, self._on_library_write)
+        framework.observe(self.on[LIBRARY_INFO_ACTION].action, self._on_library_info)
+        self.unit.set_ports(8083)
+
+    def _on_collect_status(self, event: ops.CollectStatusEvent) -> None:
+        logger.debug("_on_collect_status")
+        event.add_status(ops.ActiveStatus())
+        _, status = self._get_library_write_behaviour()
+        event.add_status(status)
+
+    def _on_pebble_ready(self, event: ops.PebbleReadyEvent) -> None:
+        """Define and start the workload using the Pebble API."""
+        container = event.workload
+        container.add_layer(SERVICE_NAME, {**self.get_pebble_layer()}, combine=True)
+        container.replan()
+
+    def _on_storage_attached(self, event: ops.StorageAttachedEvent) -> None:
+        self._push_library_to_storage()
+
+    def _on_config_changed(self, event: ops.ConfigChangedEvent):
+        """Don't do anything! User can run library-write action if needed.
```

Comment:
Does `collect-status` get triggered on every event, regardless of whether anything is observed? If so, then this can be removed. It's really just there because it used to do something, and I kept it around in case it started doing something again.